Fixes #23236 - fix for choria puppetrun on ubuntu

matt-cahill opened the pull request:
Choria's (https://github.com/choria-io/mcollective-choria)
mcollective ssl auth plugin looks for authentication keys
in the running user's home directory.

When executed via `sudo -u` '~' is expanded to
the running user's home directory rather than
the impersonated user causing an authentication error.

e.g.
```
sysadmin@puppet:/$ sudo -u foreman-proxy mco find

The find application failed to run, use -v for full 
error backtrace details: No such file or directory 
@ rb_sysopen - 
~/.puppetlabs/etc/puppet/ssl/certs/foreman-proxy.mcollective.pem
```

ekohl commented on modules/puppet_proxy_mcollective/mcollective_main.rb:
```diff
@@ -21,7 +21,8 @@ def run(nodes)
     # foreman-proxy ALL=(peadmin) NOPASSWD: /opt/puppet/bin/mco *',
     if user
-      cmd.push("-u", user)
+      homedir = File.expand_path("~#{user}")
```
Would it be good to verify ```HOME``` exists and prefer that, but use this as a fallback?

matt-cahill replied:
I think `HOME` does exist it's just the value contains the home directory for the running user not the impersonated user. I'll do some tests to confirm though.

ekohl replied:
Oh, that's a subtlety I missed. For me it does set the HOME var:
```
$ cat test.py
#!/usr/bin/env python3
import os
print(os.environ['HOME'])
$ ./test.py 
/home/ekohl
$ sudo -u root ./test.py
/root
```

There could be a different interaction in a non-interactive shell though.

matt-cahill replied:
Ok I think this is actually an ubuntu vs other distros default behaviour thing, sorry about this

```
sysadmin:~$ ruby -e 'puts File.expand_path("~"); puts ENV.fetch("HOME")'
/home/sysadmin
/home/sysadmin
sysadmin:~$ sudo -u root ruby -e 'puts File.expand_path("~"); puts ENV.fetch("HOME")'
/home/sysadmin
/home/sysadmin
```
(same in python)

However setting `Defaults        env_keep -= "HOME"` in `/etc/sudoers` gives the behaviour you are seeing so this PR isn't needed I don't think. Maybe just a note in the docs?

ekohl replied:
I figured it would be something like that. My sudoers (Fedora 27) has `Defaults env_reset` and then doesn't list `HOME` in `env_keep`.

Which version of Ubuntu are you on?
Is this the default config? If so, it'd be nice to implement a solution so it works out of the box.

matt-cahill replied:
They're stock Ubuntu 16.04.2 LTS with basically just the foreman puppet::server and foreman-proxy puppet modules applied. Perhaps adding another section to the foreman-proxy sudoers.d file (or a separate file) is better?
It would also mean foreman-proxy could run `mco puppet runonce` out of the box too.

ekohl commented:
I just saw on centos-devel that someone referred to `sudo -H` or `sudo --set-home`:
> Request that the security policy set the HOME environment variable to the home directory specified by the target user's password database entry.  Depending on the policy, this may be the default behavior.

I think this is exactly what we want.

matt-cahill commented:
Yeah that looks good, I'll edit the PR.
```
sudo -Hu root ruby -e 'puts File.expand_path("~"); puts ENV.fetch("HOME")'
/root
/root
```

mmoll commented:
[test default]

mmoll commented:
[test]

mmoll commented:
@matt-cahill please check the test failure and update the expectation accordingly :bowtie:

Commit:
Fixes #23236 - fix for choria puppetrun on ubuntu

Choria's (https://github.com/choria-io/mcollective-choria)
mcollective ssl auth plugin looks for authentication keys
in the running user's home directory.

When executed via `sudo -u` '~' is expanded to
the running user's home directory rather than
the impersonated user causing an authentication error.

e.g.
```
sysadmin@puppet:/$ sudo -u foreman-proxy mco find

The find application failed to run, use -v for full
error backtrace details: No such file or directory
@ rb_sysopen -
~/.puppetlabs/etc/puppet/ssl/certs/foreman-proxy.mcollective.pem
```

This change forces sudo to pick up the home directory
of the impersonated user from the passwd database
rather than $HOME.

matt-cahill commented:
sorry about that @mmoll, that should be in spec now

mmoll commented:
@matt-cahill I don't see the needed change, which would be in `test/puppet/mcollective_test.rb` line 23

matt-cahill committed:
Fixes #23236 - update mcollective_test.rb

matt-cahill commented:
sorry again, I was only seeing the prprocessor failures, I've made the test change too now.

mmoll commented:
[test]

ekohl commented:
ok to test

mmoll commented:
:green_heart:

mmoll commented:
[test default]

mmoll commented:
anyway, this was :green_heart:

mmoll commented:
merged, thanks @matt-cahill!

matt-cahill commented:
Thanks @mmoll :)
